### PR TITLE
Database layer

### PR DESCRIPTION
**Pull request by lohanidamodar**
## What does this PR do?

- Adds database layer to the database service above collection

## Test Plan

- Relevant tests are added

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

### CODE REVIEW

File: app/config/collections.php [hunk 1]
```diff
@@ -119,13 +177,24 @@
...
+                '$id' => 'databaseId',
+                'type' => Database::VAR_STRING,
+                'format' => '',
+                'size' => Database::LENGTH_KEY,
+                'signed' => true,
+                'required' => true,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
```
Comment (TorstenDittmann) [inline app/config/collections.php:190]: we need an index for this addition 👍🏻

File: app/controllers/api/database.php [hunk 2]
```diff
@@ -134,18 +147,353 @@ function createAttribute(string $collectionId, Document $attribute, Response $re
...
+    ->inject('events')
+    ->action(function (string $databaseId, string $name, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events) {
+
+        $databaseId = $databaseId == 'unique()' ? $dbForProject->getId() : $databaseId;
+
+        try {
+            $dbForProject->createDocument('databases', new Document([
+                '$id' => $databaseId,
+                '$read' => $read ?? [], // Collection permissions for collection documents (based on permission model)
+                '$write' => $write ?? [], // Collection permissions for collection documents (based on permission model)
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/controllers/api/database.php:177]: probably 👍🏻

File: app/controllers/api/database.php [hunk 3]
```diff
@@ -579,9 +754,19 @@ function createAttribute(string $collectionId, Document $attribute, Response $re
     ->inject('audits')
     ->inject('usage')
     ->inject('events')
-    ->action(function (string $collectionId, string $name, string $permission, ?array $read, ?array $write, bool $enabled, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events) {
+    ->action(function (string $databaseId, string $collectionId, string $name, string $permission, ?array $read, ?array $write, bool $enabled, Response $response, Database $dbForProject, EventAudit $audits, Stats $usage, Event $events) {
+        /** @var Appwrite\Utopia\Response $response */
+        /** @var Utopia\Database\Database $dbForProject */
+        /** @var Appwrite\Event\Audit $audits */
+        /** @var Appwrite\Stats\Stats $usage */
+        /** @var Appwrite\Event\Event $events */
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/controllers/api/database.php:762]: Ah yes!

File: app/init.php [hunk 4]
```diff
@@ -254,10 +255,12 @@ function (mixed $value) {
         return null;
     },
     function (mixed $value, Document $document, Database $database) {
+        $databaseId = str_replace('database_', '', $document->getAttribute('$collection'));
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/init.php:258]: so the collection's $collection will be `database_DATABASE_INTERNAL_ID` so getting the internal ID from the collection id

[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/init.php:258]: Should I instead add a databaseId attribute?

File: app/config/collections.php [hunk 5]
```diff
@@ -119,13 +155,24 @@
         '$id' => 'attributes',
         'name' => 'Attributes',
         'attributes' => [
+            [
+                '$id' => 'databaseId',
+                'type' => Database::VAR_STRING,
```
Comment (fogelito) [inline app/config/collections.php:160]: This supposed to be integer

File: app/config/collections.php [hunk 6]
```diff
@@ -263,13 +310,24 @@
         '$id' => 'indexes',
         'name' => 'Indexes',
         'attributes' => [
+            [
+                '$id' => 'databaseId',
+                'type' => Database::VAR_STRING,
```
Comment (fogelito) [inline app/config/collections.php:315]: This supposed to be Integer field

File: app/config/collections.php [hunk 7]
```diff
@@ -249,11 +296,11 @@
         ],
         'indexes' => [
             [
-                '$id' => '_key_collection',
+                '$id' => '_key_db_collection',
                 'type' => Database::INDEX_KEY,
-                'attributes' => ['collectionId'],
-                'lengths' => [Database::LENGTH_KEY],
-                'orders' => [Database::ORDER_ASC],
+                'attributes' => ['collectionId', 'databaseId'],
```
Comment (fogelito) [inline app/config/collections.php:301]: If we there is a query of databaseId without collectionId so better ['databaseId', 'collectionId']

File: app/config/collections.php [hunk 8]
```diff
@@ -263,13 +332,24 @@
...
+                '$id' => 'databaseId',
+                'type' => Database::VAR_STRING,
+                'format' => '',
+                'size' => Database::LENGTH_KEY,
+                'signed' => true,
+                'required' => true,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
```
[the remark this answers is not in the document]
Comment (fogelito) [reply app/config/collections.php:323]: Is it possible to use Internal Ids?

[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/config/collections.php:323]: I'll check if it is.

**Comment by lohanidamodar** [comment]
@gewenyu99 Waiting for documentation changes, rest should be done!

**Review by TorstenDittmann — APPROVED** [review]

**Comment by lohanidamodar** [comment]
@TorstenDittmann oh, wow. This was my best huge PR till now. Thanks 🙏🏻

**Comment by Meldiron** [comment]
@lohanidamodar Didn't review yet, wanted to play around first. I made a project, database, and attribute (string 255, required). It is stuck at processing, tho there are no errors in the worker. Can you please try to reproduce?

**Review by Meldiron — APPROVED** [review]
Looks good, but was really hard to review. I will try to review it again tomorrow with new set of eyes 👀

Either way, we will need to heavily test this before release.

**Comment by lohanidamodar** [comment]
> @lohanidamodar Didn't review yet, wanted to play around first. I made a project, database, and attribute (string 255, required). It is stuck at processing, tho there are no errors in the worker. Can you please try to reproduce?

May be something to do with cache or anything? I've not been able to reproduce

**Pull request merged**